dds/idl/be_global.cpp

```diff
@@ -602,3 +613,147 @@ BE_GlobalData::get_include_block(BE_GlobalData::stream_enum_t which)
 }
+
+void
+BE_GlobalData::cache_annotations()
+{
+  if (idl_global->idl_version_ >= IDL_VERSION_4) {
+    UTL_Scope* root = idl_global->scopes().bottom();
+    topic_annotation_ =
+      AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@topic"));
+    key_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@key"));
+    nested_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@nested"));
+    default_nested_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@default_nested"));
+ }
+}
+
+/**
+ * @brief determines if we should treat a given node as a topic type. This could
+ * mean that it is explicitly a topic type or that @default_nested(FALSE) or
+ * --default_nested = false
+ * @param node the incomming node
+ * @return if type support should be generated
+ * @see BE_GlobalData::is_topic_type(AST_Decl* node)
+ * @see BE_GlobalData::is_default_nested(AST_Decl* node)
+ * @author ceneblock
+ */
+bool BE_GlobalData::treat_as_topic(AST_Decl *node)
+{
+  return is_topic_type(node) || !is_nested_type(node);
+}
+
+/**
+ * @brief determins if a given node is nested by default or not. This may be set
+ * by the module level of @default_nesed or through the --default_nested=
+ * @param node the node to be inspected
+ * @return if the node is default nested or not
+ * @note false means it should be considered nested.
+ * @author ceneblock
+ */
+bool
+BE_GlobalData::is_default_nested(AST_Decl* node)
+{
+  if(node->defined_in()){
+
+    ///recurse to the top
+    bool default_nested_apply_value = is_default_nested(dynamic_cast<AST_Decl *>(node -> defined_in()));
+
+    ///check if we have a default value
+    AST_Annotation_Appl *default_nested_apply = dynamic_cast<AST_Decl *>(node -> defined_in()) -> annotations().find("::@default_nested");
+
+    ///if we have a default value, then overwrite the parent's.
+    if(default_nested_apply){
+      default_nested_apply_value = AST_Annotation_Member::narrow_from_decl ((*default_nested_apply)["value"]) -> value ()->ev ()->u.bval;
+    }
+    return default_nested_apply_value;
+  }
+  return default_nested_;
+}
+
+bool
+BE_GlobalData::is_topic_type(AST_Decl* node)
+{
+  if (node) {
+    if (node->node_type() == AST_Decl::NT_struct) {
+      return node->annotations().find(topic_annotation_);
+    } else if (node->node_type() == AST_Decl::NT_union) {
+      return node->annotations().find(topic_annotation_);
+    } else if (node->node_type() == AST_Decl::NT_typedef) {
+      AST_Type* type = dynamic_cast<AST_Type*>(node)->unaliased_type();
+      if (type->node_type() == AST_Decl::NT_struct
+          || type->node_type() == AST_Decl::NT_union) {
+        return node->annotations().find(topic_annotation_);
+      }
+    }
+  }
+  return false;
+}
+
+bool
+BE_GlobalData::is_key(AST_Field* node)
+{
+  // Check for @key
+  return node && node->annotations().find(key_annotation_);
```

Comment:
Does find return NULL on failure or something?

Reply:
It returns a pointer to the specific annotation object or null if the node isn't annotationed with that annotation or the annotation pointer argument is null.